handler/ratelimit/limiter.go

```diff
@@ -0,0 +1,170 @@
+package ratelimit
+
+import (
+	"net/http"
+	"runtime/debug"
+	"sync"
+	"time"
+
+	"github.com/avenga/couper/errors"
+)
+
+type Limiter struct {
+	check     chan *slowTrip
+	limits    RateLimits
+	mu        sync.RWMutex
+	transport http.RoundTripper
+}
+
+type slowTrip struct {
+	err    error
+	out    chan *slowTrip
+	quitCh <-chan struct{}
+	req    *http.Request
+	res    *http.Response
+}
+
+func NewLimiter(transport http.RoundTripper, limits RateLimits) *Limiter {
+	if len(limits) == 0 {
+		return nil
+	}
+
+	limiter := &Limiter{
+		check:     make(chan *slowTrip),
+		limits:    limits,
+		transport: transport,
+	}
+
+	for _, rl := range limits {
+		// Init the start of a period.
+		rl.periodStart = time.Now()
+	}
+
+	go limiter.slowTripper()
+
+	return limiter
+}
+
+func (l *Limiter) RoundTrip(req *http.Request) (*http.Response, error) {
+	outCh := make(chan *slowTrip)
+
+	trip := &slowTrip{
+		out:    outCh,
+		quitCh: l.limits[0].quitCh,
+		req:    req,
+	}
+
+	select {
+	case l.check <- trip:
+	case <-req.Context().Done():
+		return nil, req.Context().Err()
+	}
+
+	trip = <-outCh
+
+	return trip.res, trip.err
+}
+
+func (l *Limiter) slowTripper() {
+	defer func() {
+		if rc := recover(); rc != nil {
+			l.limits[0].logger.WithField("panic", string(debug.Stack())).Panic(rc)
+		}
+	}()
+
+	for {
+		select {
+		case <-l.limits[0].quitCh:
+			return
+		case trip := <-l.check:
+			select {
+			case <-trip.req.Context().Done():
+				// The request was canceled while in the queue.
+				trip.err = trip.req.Context().Err()
+				trip.out <- trip
+
+				// Do not sleep for X canceled requests.
+				continue
+			default:
+			}
+
+			l.mu.Lock()
+
+			if mode, timeToWait := l.checkCapacity(); mode == modeBlock && timeToWait > 0 {
+				// We do not wait, we want block directly.
+				trip.err = errors.BetaBackendRateLimitExceeded
+				trip.out <- trip
+
+				l.mu.Unlock()
+			} else {
+				select {
+				// Noop if 'timeToWait' is 0.
+				case <-time.After(timeToWait):
+				case <-trip.req.Context().Done():
+					// The request was canceled while in the queue.
+					trip.err = trip.req.Context().Err()
+					trip.out <- trip
+
+					// Do not sleep for X canceled requests.
+					continue
+				}
+
+				l.countRequest()
+
+				l.mu.Unlock()
```

Comment:
Since both cases have this `l.mu.Unlock`, move it down (after the `if` - `else`)?

Reply:
No, in else-case we have to unlock bevore `go func()`